dvc/cache/base.py

```diff
@@ -141,25 +141,30 @@ def _do_link(self, from_info, to_info, link_method):
             "Created '%s': %s -> %s", self.cache_types[0], from_info, to_info,
         )
 
+    def add(self, path_info, tree, hash_info):
+        if not self.changed_cache_file(hash_info):
+            return
+
+        cache_info = self.tree.hash_to_path_info(hash_info.value)
+        # using our makedirs to create dirs with proper permissions
+        self.makedirs(cache_info.parent)
+        if isinstance(tree, type(self.tree)):
+            self.tree.move(path_info, cache_info)
+        else:
+            with tree.open(path_info, mode="rb") as fobj:
+                self.tree.upload_fobj(fobj, cache_info)
+            tree.state.save(path_info, hash_info)
+        self.protect(cache_info)
+        self.tree.state.save(cache_info, hash_info)
+
     def _save_file(self, path_info, tree, hash_info, **kwargs):
         assert hash_info
 
-        cache_info = self.tree.hash_to_path_info(hash_info.value)
-        if self.changed_cache(hash_info):
```

Comment:
This `cached_cache` was costing us one `isdir()` call :man_facepalming: :man_facepalming: :man_facepalming: Should've used `changed_cache_file` all along.